### PR TITLE
Ft25 create fairgbm classifier

### LINKED ISSUE #25: Create FairGBM aliases for client-facing Python classes (e.g., `FairGBMClassifier`)
Currently the lightgbm scikit-learn API has a `LGBMClassifier`.

We would like to have an equivalent `FairGBMClassifier` that enforces the use of fairness (constraints).

This would simply be a subclass of LGBMClassifier that hard-codes the objective function to the `constrained_cross_entropy` (and perhaps some other kwargs should be enforced as well).

Also, the lightgbm package also provides another API (non sklearn), can we create a FairGBM alias for that API as well?

### PR DESCRIPTION
**Pull request by garsenio**
Closes https://github.com/feedzai/fairgbm/issues/25
Created new tests for the sklearn fairgbm api. 
Uncommented most engine tests.

### CODE REVIEW

File: python-package/fairgbm/sklearn.py [hunk 1]
```diff
@@ -1033,3 +1199,45 @@ def fit(self, X, y,
...
+                         "constrained_cross_entropy", class_weight, min_split_gain, 
+                         min_child_weight, min_child_samples, subsample, 
+                         subsample_freq, colsample_bytree, reg_alpha, 
+                         reg_lambda, random_state, n_jobs, 
+                         silent, importance_type, multipler_learning_rate=multipler_learning_rate,
+                         constraint_type=constraint_type, global_constraint_type=global_constraint_type,
+                         global_target_fpr=global_target_fpr, global_target_fnr=global_target_fnr)
+
+    def fit(self, X, y,
+            constraint_group,
```
Comment (AndreFCruz) [inline python-package/fairgbm/sklearn.py:1234]: `constraint_group` should be a madatory key-word-only argument. I think this is warranted as the standard sklearn API is wide-spread (`fit(X, y)`), and the extra-argument may be hard to decipher without more context (context -> explicitly stating the key-word).

See: https://peps.python.org/pep-3102/#:~:text=Keyword-only%20arguments%20are%20not,therefore%20%27required%20keyword%27%20arguments.

File: tests/python_package_test/test_engine.py [hunk 2]
```diff
@@ -1,1588 +1,1588 @@
...
-#     lgb_train = lgb.Dataset(X, y, free_raw_data=False)
-#     init_gbm = lgb.train(params, lgb_train, num_boost_round=5)
-#     init_gbm_2 = lgb.train(params, lgb_train, num_boost_round=5, init_model=init_gbm)
-#     init_gbm_3 = lgb.train(params, lgb_train, num_boost_round=5, init_model=init_gbm_2)
-#     gbm = lgb.train(params, lgb_train, num_boost_round=5, init_model=init_gbm_3)
-#     assert gbm.current_iteration() == 20
-
-
-# def test_continue_train_dart():
+# coding: utf-8
```
Comment (AndreFCruz) [inline tests/python_package_test/test_engine.py:1]: Why was this file commented out? Is this some mishap from a previous PR?

File: python-package/fairgbm/sklearn.py (hunk 1 above)
Comment (garsenio) [reply python-package/fairgbm/sklearn.py:1234]: Good improvement, will change :+1:

File: tests/python_package_test/test_engine.py (hunk 2 above)
Comment (garsenio) [reply tests/python_package_test/test_engine.py:1]: Your question is why was test_engine.py commented before this PR? It was failing and the reason was not clear. I commented it and put it on backlog. Now, somehow, it works.

File: python-package/fairgbm/sklearn.py [hunk 3]
```diff
@@ -1033,3 +1199,45 @@ def fit(self, X, y,
...
+                         learning_rate, n_estimators, subsample_for_bin, 
+                         "constrained_cross_entropy", class_weight, min_split_gain, 
+                         min_child_weight, min_child_samples, subsample, 
+                         subsample_freq, colsample_bytree, reg_alpha, 
+                         reg_lambda, random_state, n_jobs, 
+                         silent, importance_type, multipler_learning_rate=multipler_learning_rate,
+                         constraint_type=constraint_type, global_constraint_type=global_constraint_type,
+                         global_target_fpr=global_target_fpr, global_target_fnr=global_target_fnr)
+
+    def fit(self, X, y, *,
```
Comment (AndreFCruz) [inline python-package/fairgbm/sklearn.py:1233]: nice 👍

File: tests/python_package_test/test_sklearn.py [hunk 4]
```diff
@@ -79,6 +79,14 @@ def multi_error(y_true, y_pred):
 def multi_logloss(y_true, y_pred):
     return np.mean([-math.log(y_pred[i][y]) for i, y in enumerate(y_true)])
 
+def test_binary_fairgbm():
```
Comment (AndreFCruz) [inline tests/python_package_test/test_sklearn.py:83]: can we add a short test to assert that calling fairgbm.fit without the constraint_group fails? (should be a TypeError)

**Comment by AndreFCruz** [comment]
Should we add a TODO on the C++ metric constructor to add a constrained_cross_entropy metric? It works fine without it but it still tries to fetch such a metric, perhaps it would be useful to create one in the future.

Should we create an issue and link it in that TODO ?

**Comment by GoncaloArs** [comment]
Added new issue https://github.com/feedzai/fairgbm/issues/29

**Comment by garsenio** [comment]
> Should we add a TODO on the C++ metric constructor to add a constrained_cross_entropy metric? It works fine without it but it still tries to fetch such a metric, perhaps it would be useful to create one in the future.
> 
> Should we create an issue and link it in that TODO ?

Added new issue https://github.com/feedzai/fairgbm/issues/29

**Review by AndreFCruz — APPROVED** [review]
good job :+1: 

ready to merge imo

**Review by AndreFCruz — CHANGES_REQUESTED** [review]
I'll address a few minor things:
- arguments should be passed from FairGBM constructor to LGBMModel as key-words instead of positional arguments (less error prone)
- **kwargs passed to FairGBM constructor should be passed on to the super constructor

File: python-package/fairgbm/sklearn.py [hunk 5]
```diff
@@ -1033,3 +1199,51 @@ def fit(self, X, y,
...
                    f"{' ':12}The evaluation positions of the specified metric.\n"
                    f"{' ':8}{_early_stop}{_after_early_stop}")
+
+
+class FairGBMClassifier(LGBMClassifier):
+    """Helper wrapper to use FairGBM with an sklearn-like API."""
+
+    FAIRGBM_OBJECTIVE = "constrained_cross_entropy"
+
+    def __init__(self, boosting_type='gbdt', num_leaves=31,
```
Comment (AndreFCruz) [inline python-package/fairgbm/sklearn.py:1209]: Question, should we re-define the default values of LGBM key-word args here?

Perhaps it would be better to just list FairGBM-only args here and leave the rest to a `**kwargs` wildcard. That way all default values are maintained for common LGBM arguments.

File: python-package/fairgbm/sklearn.py [hunk 6]
```diff
@@ -1033,3 +1199,51 @@ def fit(self, X, y,
...
+
+    def __init__(self, boosting_type='gbdt', num_leaves=31, 
+                 max_depth=-1, learning_rate=0.1, n_estimators=100, 
+                 subsample_for_bin=200000, class_weight=None, 
+                 min_split_gain=0, min_child_weight=0.001, min_child_samples=20, 
+                 subsample=1, subsample_freq=0, colsample_bytree=1, 
+                 reg_alpha=0, reg_lambda=0, random_state=None, 
+                 n_jobs=-1, silent=True, importance_type='split',
+                 multipler_learning_rate=50_000, constraint_type="FPR", 
+                 global_constraint_type="FPR, FNR", global_target_fpr=0.05,
```
Comment (AndreFCruz) [inline python-package/fairgbm/sklearn.py]: Should we set these defaults for the open-source package? @saleiro 

I think a lot of people will use the default values and these are somewhat specific to our projects.

**Review by AndreFCruz — CHANGES_REQUESTED** [review]
Raised a few points to be discussed

**Comment by garsenio** [comment]
> Raised a few points to be discussed

Since these points are not critical can we open issues with these discussions and merge this PR? @AndreFCruz

File: python-package/fairgbm/sklearn.py (hunk 5 above)
Comment (AndreFCruz) [reply python-package/fairgbm/sklearn.py:1209]: EDIT: all sklearn-api classifiers must declare all init arguments explicitly as key-word args; therefore this isn't valid

File: python-package/fairgbm/sklearn.py (hunk 6 above)
Comment (AndreFCruz) [reply python-package/fairgbm/sklearn.py]: Moved to #31

**Review by AndreFCruz — APPROVED** [review]
